### ISSUE TITLE
Cryptomatte not working with DWA compression

### ISSUE THREAD

**Issue by sharktacos** [Psyop/Cryptomatte#35]
Cryptomatte works fine on a multichannel EXR with ZIP scanline compression, but not with DWAA or DWAB compression. That is, it is not possible to select the objects in Nuke with the Cryptomatte gizmo when the image is rendered with DWA compression. Experiencing this both with Arnold and Vray renders.

**Comment by anderslanglands** [comment]
DWA compression is lossy. Cryptomatte relies on exact values in the images,
so lossy compression will break it.

**Comment by dekekincaid** [comment]
DWA is only lossy on channels named:
R, G, B, RY, BY, Y (case sensitive). All other channel names are treated as lossless.

**Comment by anderslanglands** [comment]
Yeah was just reading the source code. Yay for magic constants. So...
everyone has to change to outputting .xyz instead of .rgb now?

**Comment by sharktacos** [comment]
x y z should work, just be careful of Y (capitol Y is lossy). I believe if they were named r g b (lowercase) that would work too. They just can't be named R G B (capitol letters).

**Comment by anderslanglands** [comment]
rgb gets lossy compression as well. Full list is here in function
initializeDefaultChannelRules():
https://github.com/openexr/openexr/blob/develop/OpenEXR/IlmImf/ImfDwaCompressor.cpp

Would be nice if they'd allowed for customising it via header metadata or
something rather than just hard-coding it...

**Comment by vkoylazov** [comment]
This will be fixed in a V-Ray nightly soon. To fix it we patched the OpenEXR2 library to allow us to explicitly specify if a channel should be DWA compressed.

Best regards,
Vlado
________________________________________

**Comment by jonahfriedman** [comment]
@vkoylazov Very cool. Are you compressing the coverages? Roughly what kind of file size improvements are you seeing with DWA Cryptomattes?

**Comment by jonahfriedman** [comment]
Another observation: While you are required to have the same compression for the whole file, you are allowed to have different channels with different data types per channel. ([EXR technical introduction](http://www.openexr.com/TechnicalIntroduction.pdf), page 7), and having the coverages at half I think would always make sense, maybe even with lossy compression. 

We'd have to test pretty carefully that images with RGBA set to different data types don't confuse the various applications, but that's another thing to try.

**Comment by dekekincaid** [comment]
rgb doesn’t get lossy.  If you look at the source that is only for openexr 1.0 which was before dwa shipped to the public. 2.0 is only for capital RGBYGYRY.
